R/group_apply.R

```diff
@@ -0,0 +1,133 @@
+#' supplied dataset, enabling the separate analysis of occurrences or taxa from
+#' different time intervals, spatial regions, or trait values. The function
+#' serves as a wrapper around `palaeoverse` functions. Other functions which
+#' can be applied to a `data.frame` (e.g. `nrow`, `ncol`, `lengths`,
+#' `unique`) may also be used.
+#' \cr
+#' \cr
+#' All `palaeoverse` functions which require a dataframe input can be used in
+#' conjunction with the `group_apply` function. However, this is unnecessary
+#' for many functions (e.g. \code{\link{bin_time}}) as groups do not need to
+#' be partitioned before binning. This list provides
+#' users with `palaeoverse` functions that might be interesting to apply across
+#' group(s):
+#' - \code{\link{tax_unique}}: return the number of unique taxa per grouping
+#' variable.
+#' - \code{\link{tax_range_time}}: return the temporal range of taxa per
+#' grouping variable.
+#' - \code{\link{tax_range_geo}}: return the geographic range of taxa per
+#' grouping variable.
+#' - \code{\link{tax_check}}: return potential spelling variations of the
+#' same taxon per grouping variable. Note: `verbose` needs to be set to FALSE.
+#'
+#'
+#' @section Developer(s):
+#' Lewis A. Jones & William Gearty
+#' @section Reviewer(s):
+#' Kilian Eichenseer & Bethany Allen
+#' @importFrom stats as.formula
+#' @examples
+#' # Examples
+#' # Get tetrapods data
+#' occdf <- tetrapods
+#' # Remove NA data
+#' occdf <- subset(occdf, !is.na(genus))
+#' # Count number of occurrences from each country
+#' ex1 <- group_apply(occdf = occdf, group = "cc", fun = nrow)
+#' # Temporal range of taxa per country with default arguments
+#' ex2 <- group_apply(occdf = occdf, group = "cc", fun = tax_range_time)
+#' # Temporal range of taxa per country with updated arguments
+#' ex3 <- group_apply(occdf = occdf,
+#'                    group = c("cc"),
+#'                    fun = tax_range_time,
+#'                    by = "LAD") # Order by LAD (default: "FAD")
+#' # Use multiple variables (number of occurrences per collection and formation)
+#' ex4 <- group_apply(occdf = occdf,
+#'                    group = c("collection_no", "formation"),
+#'                    fun = nrow)
+#' # Compute counts of occurrences per latitudinal bin
+#' # Set up lat bins
+#' bins <- lat_bins()
+#' # bin occurrences
+#' occdf <- bin_lat(occdf = tetrapods, bins = bins)
+#' # Calculate number of occurrences per bin
+#' ex5 <- group_apply(occdf = occdf, group = "lat_bin", fun = nrow)
+#' @export
+group_apply <- function(occdf, group, fun, ...) {
+
+  # Handle errors
+  if (!is.data.frame(occdf)) {
+    stop("`occdf` should be a dataframe")
+  }
+
+  if (!any(group %in% colnames(occdf))) {
+    stop("Supplied `group` is not a named column in `occdf`")
+  }
+
+  if (!is.function(fun)) {
+    stop("Supplied `fun` is not a function")
+  }
+
+  supp_args <- list(...)
+  indx <- which(!(names(supp_args) %in% names(formals(fun))))
+  if (length(indx) > 1) {
```

Comment:
```suggestion
  if (length(indx) > 0) {
```
Shouldn't this be 0 or >= 1 to detect any additional arguments?